# Peer-graded Assignment: Segmenting and Clustering Neighborhoods in Toronto
## Part 3: Clustering the neighborhoods in Toronto
### Install and import necessary packages

In [1]:
!pip install bs4
!pip install geopy
import pandas as pd
import numpy as np
from bs4 import BeautifulSoup
import requests
import json
from geopy.geocoders import Nominatim
import folium
from sklearn.cluster import KMeans
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

### Set up the dataframes as they were at the end of Part 2

In [2]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
html = requests.get(url)
table = BeautifulSoup(html.text, 'html.parser').find('table')  #Finds the first table in the page
table_rows = table.find_all('tr')                              #Finds all <tr> tags and returns a list of them
data = []
for row in table_rows:                               #Iterate through all rows
    tds = row.find_all('td')                         #Find all <td> tags
    new_row = [td.text.strip() for td in tds]        #Create a list of the td elements, while removing the "\n" from the end
    if len(new_row):
        data.append(new_row)
df = pd.DataFrame(data, columns=["Postal code", "Borough", "Neighborhood"])
df = df.drop(df[df["Borough"] == "Not assigned"].index)
df["Neighborhood"] = df["Neighborhood"].str.replace(" / ", ", ")
df["Neighborhood"] = df["Neighborhood"].str.replace("\\n", "")
df.reset_index(inplace = True, drop = True)
coords = pd.read_csv("https://cocl.us/Geospatial_data")
toronto = df.join(other=coords.set_index("Postal Code"), on="Postal code")

In [3]:
print(toronto.shape)
toronto.head()

(103, 5)


Postal code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

### Using the geopy library we determine the coordinates of Toronto

In [4]:
geolocator = Nominatim(user_agent="toronto_cluster")
location = geolocator.geocode("Toronto, Ontario, Canada")
toronto_lat = location.latitude
toronto_long = location.longitude
print(toronto_lat, toronto_long)

43.6534817 -79.3839347


### Visualize Toronto's neighborhoods on a map

In [5]:
map = folium.Map(location=[toronto_lat, toronto_long], zoom_start=10)

for lat, lng, borough, neighborhood in zip(toronto['Latitude'], toronto['Longitude'], toronto['Borough'], toronto['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map)  
map

### Set up Fourquare credentials

In [6]:
CLIENT_ID = 'ZBC2EWETE1R2DGRBAWJ1YGOOUBCTXCOZTJS4JHQ0V1MBJPJU' # your Foursquare ID
CLIENT_SECRET = 'UGTY5QSD0CVQQC1DFIKRZULRMNVZJSFAPKVHWKSZXR0ADP5Z' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100

### Define a function to get the category of the resulted venue. If more than one categories are given, then the function returns only the first one.

In [7]:
def get_category(row):
    category_list = row["venue.categories"]
    if len(category_list):
        return category_list[0]["name"]
    else:
        return None

### Let's explore the first neighborhood, Parkwoods.

In [9]:
radius = 700

latitude = toronto.iloc[0,3]
longitude = toronto.iloc[0,4]

url = url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET,
            VERSION,
            latitude,
            longitude,
            radius,
            LIMIT)

results = requests.get(url).json()
venues = results['response']['groups'][0]['items']
df_venues = pd.json_normalize(venues)
df_venues = df_venues[["venue.name", "venue.location.lat", "venue.location.lng", "venue.categories"]]
df_venues["venue.categories"] = df_venues.apply(get_category, axis=1)
df_venues.columns = [col.split('.')[-1].capitalize() for col in df_venues.columns]
print(df_venues.shape)
df_venues

(5, 4)


Name        Lat        Lng         Categories
0     Brookbanks Park  43.751976 -79.332140               Park
1            PetSmart  43.748639 -79.333488          Pet Store
2       Variety Store  43.751974 -79.333114  Food & Drink Shop
3                Joey  43.753441 -79.321640       Burger Joint
4  Three Valleys Park  43.751195 -79.337356               Park

With a radius of 500 meters, there are only 2 results in Parkwoods. Even with an extended radius of 700 we have found only 5 venues. So in the code below, I will extend the radius with 200 meters if the number of results is less than 10, but it is maximised at 900.

I create an empty dataframe, then append each venue one-by-one. The resulted dataframe has only 5 columns as visible below.

In [10]:
toronto_venues = pd.DataFrame(columns=["Neighborhood", "Venue name", "Venue Latitude", "Venue Longitude", "Category"])

def set_url(lat, lng, radius):
    return 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET,
            VERSION,
            lat,
            lng,
            radius,
            LIMIT)

for i, row in toronto.iterrows():
    #time.sleep(0.1)
    radius = 500
    lat = row["Latitude"]
    lng = row["Longitude"]
    nbhood = row["Neighborhood"]
    print(i, nbhood, "...", end='')
    while True:
        url = set_url(lat, lng, radius)
        results = requests.get(url).json()
        venues = results['response']['groups'][0]['items']
        if len(venues) >= 10 or radius >= 900:
            break
        radius += 200
    if len(venues) == 0:
        print("could not find any venues in {}!".format(nbhood))
        continue
    df_venues = pd.json_normalize(venues)
    df_venues = df_venues[["venue.name", "venue.location.lat", "venue.location.lng", "venue.categories"]]
    df_venues["venue.categories"] = df_venues.apply(get_category, axis=1)
    for j, venue in df_venues.iterrows():
        toronto_venues = toronto_venues.append({
            "Neighborhood": nbhood,
            "Venue name": df_venues.loc[j, "venue.name"],
            "Venue Latitude": df_venues.loc[j, "venue.location.lat"],
            "Venue Longitude": df_venues.loc[j, "venue.location.lng"],
            "Category": df_venues.loc[j, "venue.categories"]
        }, ignore_index=True)
    print("  {} venues found!".format(len(venues)))
    
print(toronto_venues.shape)
toronto_venues.head()

0 Parkwoods ...  17 venues found!
1 Victoria Village ...  10 venues found!
2 Regent Park, Harbourfront ...  47 venues found!
3 Lawrence Manor, Lawrence Heights ...  14 venues found!
4 Queen's Park, Ontario Provincial Government ...  31 venues found!
5 Islington Avenue ...  10 venues found!
6 Malvern, Rouge ...  17 venues found!
7 Don Mills ...  13 venues found!
8 Parkview Hill, Woodbine Gardens ...  13 venues found!
9 Garden District, Ryerson ...  100 venues found!
10 Glencairn ...  18 venues found!
11 West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale ...  17 venues found!
12 Rouge Hill, Port Union, Highland Creek ...  4 venues found!
13 Don Mills ...  23 venues found!
14 Woodbine Heights ...  25 venues found!
15 St. James Town ...  86 venues found!
16 Humewood-Cedarvale ...  25 venues found!
17 Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood ...  18 venues found!
18 Guildwood, Morningside, West Hill ...  22 venues found!
19 The Beaches ...  41 venues

Neighborhood         Venue name  Venue Latitude  Venue Longitude  \
0    Parkwoods    Allwyn's Bakery       43.759840       -79.324719   
1    Parkwoods    Brookbanks Park       43.751976       -79.332140   
2    Parkwoods        Tim Hortons       43.760668       -79.326368   
3    Parkwoods                A&W       43.760643       -79.326865   
4    Parkwoods  Bruno's valu-mart       43.746143       -79.324630   

               Category  
0  Caribbean Restaurant  
1                  Park  
2                  Café  
3  Fast Food Restaurant  
4         Grocery Store

Cool! So, we have found 2738 venues in Toronto. But there are not any venues in Upper Rouge. That may cause problems later!

And how many categories?

In [11]:
print('There are {} unique categories.'.format(len(toronto_venues["Category"].unique())))

There are 284 unique categories.


I generate one-hot-encoding

In [12]:
toronto_onehot = pd.get_dummies(toronto_venues[["Category"]], prefix="", prefix_sep="")
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]
toronto_onehot.head()

Yoga Studio  Accessories Store  Airport  Airport Food Court  Airport Gate  \
0            0                  0        0                   0             0   
1            0                  0        0                   0             0   
2            0                  0        0                   0             0   
3            0                  0        0                   0             0   
4            0                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Arts & Crafts Store  Asian Restaurant  \
0             0         0            0                    0                 0   
1             0         0            0                    0                 0   
2             0         0            0                    0                 0   
3             0         0            0                    0                 0   
4             0         0            0                    0                 0   

   Athletics & Sports  Auto Garage  Auto Workshop  BBQ Joint  Baby Store  \
0                   0            0              0          0           0   
1                   0            0              0          0           0   
2                   0            0              0          0           0   
3                   0            0              0          0           0   
4                   0            0              0          0           0   

   Badminton Court  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0                0           0       0     0    0               0   
1                0           0       0     0    0               0   
2                0           0       0     0    0               0   
3                0           0       0     0    0               0   
4                0           0       0     0    0               0   

   Baseball Stadium  Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  \
0                 0                   0      0                0         0   
1                 0                   0      0                0         0   
2                 0                   0      0                0         0   
3                 0                   0      0                0         0   
4                 0                   0      0                0         0   

   Beer Store  Belgian Restaurant  Bike Shop  Bistro  Board Shop  \
0           0                   0          0       0           0   
1           0                   0          0       0           0   
2           0                   0          0       0           0   
3           0                   0          0       0           0   
4           0                   0          0       0           0   

   Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  \
0              0          0         0                     0               0   
1              0          0         0                     0               0   
2              0          0         0                     0               0   
3              0          0         0                     0               0   
4              0          0         0                     0               0   

   Brewery  Bridal Shop  Bubble Tea Shop  Building  Burger Joint  \
0        0            0                0         0             0   
1        0            0                0         0             0   
2        0            0                0         0             0   
3        0            0                0         0             0   
4        0   

### Wow! How is it possible that Yoga Studio comes to the first place? 'Neighborhood' did not added? Seems so. But look at below: We still can group the results by neighborhood. Is this a bug in pandas? Just let me know if you have any idea :)

In [13]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                                           Agincourt     0.000000   
1                              Alderwood, Long Branch     0.000000   
2     Bathurst Manor, Wilson Heights, Downsview North     0.000000   
3                                     Bayview Village     0.000000   
4                   Bedford Park, Lawrence Manor East     0.000000   
5                                         Berczy Park     0.000000   
6                         Birch Cliff, Cliffside West     0.000000   
7        Brockton, Parkdale Village, Exhibition Place     0.041667   
8               Business reply mail Processing CentrE     0.052632   
9   CN Tower, King and Spadina, Railway Lands, Har...     0.000000   
10                                Caledonia-Fairbanks     0.000000   
11              Canada Post Gateway Processing Centre     0.000000   
12                                          Cedarbrae     0.000000   
13                                 Central Bay Street     0.015385   
14                                           Christie     0.000000   
15                               Church and Wellesley     0.027027   
16            Clarks Corners, Tam O'Shanter, Sullivan     0.000000   
17    Cliffside, Cliffcrest, Scarborough Village West     0.000000   
18                     Commerce Court, Victoria Hotel     0.000000   
19                                         Davisville     0.000000   
20                                   Davisville North     0.000000   
21   Del Ray, Mount Dennis, Keelsdale and Silverthorn     0.000000   
22                                          Don Mills     0.000000   
23  Dorset Park, Wexford Heights, Scarborough Town...     0.000000   
24                                          Downsview     0.000000   
25                       Dufferin, Dovercourt Village     0.000000   
26                                       East Toronto     0.000000   
27  Eringate, Bloordale Gardens, Old Burnhamthorpe...     0.000000   
28                       Fairview, Henry Farm, Oriole     0.000000   
29             First Canadian Place, Underground city     0.000000   
30                           Forest Hill North & West     0.000000   
31                           Garden District, Ryerson     0.000000   
32                                          Glencairn     0.000000   
33                    Golden Mile, Clairlea, Oakridge     0.000000   
34                  Guildwood, Morningside, West Hill     0.000000   
35  Harbourfront East, Union Station, Toronto Islands     0.000000   
36                      High Park, The Junction South     0.000000   
37                                  Hillcrest Village     0.000000   
38                                      Humber Summit     0.000000   
39                                   Humberlea, Emery     0.000000   
40                                 Humewood-Cedarvale     0.000000   
41                     India Bazaar, The Beaches West     0.000000   
42                                   Islington Avenue     0.000000   
43        Kennedy Park, Ionview, East Birchmount Park     0.000000   
44          Kensington Market, Chinatown, Grange Park     0.000000   
45  Kingsview Village, St. Phillips, Martin Grove ...     0.000000   
46                   Lawrence Manor, Lawrence Heights     0.000000   
47                                      Lawrence Park     0.000000   
48                                            Leaside     0.000000   
49                           Little Portugal, Trinity     0.023256   
50                                     Malvern, Rouge     0.000000   
51  Milliken, Agincourt North, Steeles East, L'Amo...     0.000000   
52  Mimico NW, The Queensway West, South of Bloor,...     0.000000   
53                        Moore Park, Summerhill East     0.000000   
54       New Toronto, Mimico South, Humber Bay Shores     0.000000   
55           North Park, Maple Leaf Park, Upwood Park     0.000000   
56                                 North Toronto We

In [14]:
toronto_grouped.shape

(97, 284)

### Hmm... We used to have 103 Neighborhoods. Now we have only 97. Why?

In [15]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [16]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

print(neighborhoods_venues_sorted.shape)
neighborhoods_venues_sorted.head()

(97, 11)


Neighborhood 1st Most Common Venue  \
0                                        Agincourt    Chinese Restaurant   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village   Japanese Restaurant   
4                Bedford Park, Lawrence Manor East        Sandwich Place   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0         Shopping Mall                Bakery          Skating Rink   
1              Pharmacy                   Gym          Skating Rink   
2                  Bank           Pizza Place           Gas Station   
3                  Bank           Gas Station                  Café   
4    Italian Restaurant            Restaurant           Coffee Shop   

      5th Most Common Venue 6th Most Common Venue      7th Most Common Venue  \
0  Mediterranean Restaurant                  Bank             Sandwich Place   
1            Sandwich Place                  Pool                Coffee Shop   
2               Bridal Shop        Sandwich Place  Middle Eastern Restaurant   
3                Restaurant    Chinese Restaurant               Intersection   
4               Pizza Place              Pharmacy                    Butcher   

  8th Most Common Venue 9th Most Common Venue   10th Most Common Venue  
0      Sushi Restaurant           Supermarket     Caribbean Restaurant  
1                   Pub    Athletics & Sports         Doner Restaurant  
2           Supermarket                 Diner         Sushi Restaurant  
3         Shopping Mall                  Park            Grocery Store  
4                  Café                   Pub  Comfort Food Restaurant

In [17]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 1, 1, 1, 1, 1, 1], dtype=int32)

In [18]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged

Postal code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
5           M9A         Etobicoke   
6           M1B       Scarborough   
7           M3B        North York   
8           M4B         East York   
9           M5B  Downtown Toronto   
10          M6B        North York   
11          M9B         Etobicoke   
12          M1C       Scarborough   
13          M3C        North York   
14          M4C         East York   
15          M5C  Downtown Toronto   
16          M6C              York   
17          M9C         Etobicoke   
18          M1E       Scarborough   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
21          M6E              York   
22          M1G       Scarborough   
23          M4G         East York   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
26          M1H       Scarborough   
27          M2H        North York   
28          M3H        North York   
29          M4H         East York   
30          M5H  Downtown Toronto   
31          M6H      West Toronto   
32          M1J       Scarborough   
33          M2J        North York   
34          M3J        North York   
35          M4J         East York   
36          M5J  Downtown Toronto   
37          M6J      West Toronto   
38          M1K       Scarborough   
39          M2K        North York   
40          M3K        North York   
41          M4K      East Toronto   
42          M5K  Downtown Toronto   
43          M6K      West Toronto   
44          M1L       Scarborough   
45          M2L        North York   
46          M3L        North York   
47          M4L      East Toronto   
48          M5L  Downtown Toronto   
49          M6L        North York   
50          M9L        North York   
51          M1M       Scarborough   
52          M2M        North York   
53          M3M        North York   
54          M4M      East Toronto   
55          M5M        North York   
56          M6M              York   
57          M9M        North York   
58          M1N       Scarborough   
59          M2N        North York   
60          M3N        North York   
61          M4N   Central Toronto   
62          M5N   Central Toronto   
63          M6N              York   
64          M9N              York   
65          M1P       Scarborough   
66          M2P        North York   
67          M4P   Central Toronto   
68          M5P   Central Toronto   
69          M6P      West Toronto   
70          M9P         Etobicoke   
71          M1R       Scarborough   
72          M2R        North York   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
76          M7R       Mississauga   
77          M9R         Etobicoke   
78          M1S       Scarborough   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
82          M1T       Scarborough   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
85          M1V       Scarborough   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
88          M8V         Etobicoke   
89          M9V         Etobicoke   
90          M1W       Scarborough   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
93          M8W         Etobicoke   
94          M9W         Etobicoke   
95          M1X       Scarborough   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                          Neighborhood   Latitude  Longitude  \
0                                            Parkwoods  43.753259 -79.32965

Upper Rouge has a label of NaN. We were expecting some trouble here. But how does it come that we could join the 97 rows toronto_grouped table with the 103 row toronto without any problem; and 102 out of the 103 have their most common categories?

In order not to delete upper rouge, I will give it a unique label, I will assign it to a 6th cluster by hand.

In [19]:
toronto_merged.iloc[95, 5] = 5.0

And we turn the label values back to integer

In [20]:
toronto_merged["Cluster Labels"] = toronto_merged["Cluster Labels"].astype(int)
toronto_merged.head()

Postal code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656               1                  Park   
1  43.725882 -79.315572               1                  Park   
2  43.654260 -79.360636               1           Coffee Shop   
3  43.718518 -79.464763               1        Clothing Store   
4  43.662301 -79.389494               1           Coffee Shop   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Bus Stop           Coffee Shop                  Road   
1              Playground   Sporting Goods Shop                  Café   
2                     Pub                Bakery                  Park   
3  Furniture / Home Store     Accessories Store              Boutique   
4                   Diner   Fried Chicken Joint        Discount Store   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0   Caribbean Restaurant           Supermarket    Chinese Restaurant   
1  Portuguese Restaurant          Hockey Arena           Coffee Shop   
2             Restaurant        Breakfast Spot                  Café   
3  Vietnamese Restaurant           Event Space    Miscellaneous Shop   
4    Distribution Center                  Bank                   Bar   

  8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
0     Food & Drink Shop       Laundry Service               Discount Store  
1           Pizza Place          Intersection         Ethiopian Restaurant  
2    Mexican Restaurant               Theater            Health Food Store  
3           Coffee Shop             Gift Shop  Eastern European Restaurant  
4    Italian Restaurant           Yoga Studio                      Theater

In [21]:
import matplotlib.cm as cm
import matplotlib.colors as colors

In [22]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=10)

# set color scheme for the clusters
x = np.arange(kclusters+1)
ys = [i + x + (i*x)**2 for i in range(kclusters+1)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Let's see the clusters one-by-one
#### Cluster 1

In [25]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Neighborhood  1st Most Common Venue  \
8                     Parkview Hill, Woodbine Gardens            Pizza Place   
10                                          Glencairn            Pizza Place   
18                  Guildwood, Morningside, West Hill                   Park   
27                                  Hillcrest Village                   Park   
28    Bathurst Manor, Wilson Heights, Downsview North            Coffee Shop   
29                                   Thorncliffe Park      Indian Restaurant   
31                       Dufferin, Dovercourt Village               Pharmacy   
32                                Scarborough Village         Ice Cream Shop   
34                    Northwood Park, York University            Pizza Place   
39                                    Bayview Village    Japanese Restaurant   
40                                          Downsview  Vietnamese Restaurant   
45                           York Mills, Silver Hills           Tennis Court   
46                                          Downsview  Vietnamese Restaurant   
51    Cliffside, Cliffcrest, Scarborough Village West            Pizza Place   
53                                          Downsview  Vietnamese Restaurant   
56   Del Ray, Mount Dennis, Keelsdale and Silverthorn          Grocery Store   
60                                          Downsview  Vietnamese Restaurant   
64                                             Weston            Coffee Shop   
70                                          Westmount            Pizza Place   
77  Kingsview Village, St. Phillips, Martin Grove ...            Coffee Shop   
78                                          Agincourt     Chinese Restaurant   
82            Clarks Corners, Tam O'Shanter, Sullivan            Pizza Place   
85  Milliken, Agincourt North, Steeles East, L'Amo...            Pizza Place   
89  South Steeles, Silverstone, Humbergate, Jamest...          Grocery Store   
90                      Steeles West, L'Amoreaux West   Fast Food Restaurant   
93                             Alderwood, Long Branch            Pizza Place   

     2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
8       Athletics & Sports                   Pharmacy              Bus Line   
10           Grocery Store                       Park           Coffee Shop   
18    Fast Food Restaurant                Pizza Place            Sports Bar   
27                Pharmacy                Pizza Place                Bakery   
28                    Bank                Pizza Place           Gas Station   
29             Yoga Studio                       Bank           Gas Station   
31                  Bakery                Pizza Place                  Park   
32             Pizza Place                Coffee Shop        Sandwich Place   
34             Frame Store             Massage Studio  Caribbean Restaurant   
39                    Bank                Gas Station                  Café   
40           Grocery Store                       Park           Gas Station   
45               Cafeteria                       Pool         Deli / Bodega   
46           Grocery Store                       Park           Gas Station   
51          Ice Cream Shop             Hardware Store          Burger Joint   
53           Grocery Store                       Park           Gas Station   
56  Furniture / Home Store             Sandwich Place             Wine Shop   
60           Grocery Store                       Park           Gas Station   
64           Train Station             Breakfast Spot                 Diner   
70             Coffee Shop  Middle Eastern Restaurant           Flea Market   
77             Gas Station              Shopping Mall              Pharmacy   
78           Shopping Mall                     Bakery          Skating Rink   
82      Italian Restaurant        Fried Chicken Joint  Fast Food Restaurant   
85      Chinese Restaurant       Fast Food Restaurant          

#### Cluster 2

In [27]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Neighborhood  \
0                                            Parkwoods   
1                                     Victoria Village   
2                            Regent Park, Harbourfront   
3                     Lawrence Manor, Lawrence Heights   
4          Queen's Park, Ontario Provincial Government   
6                                       Malvern, Rouge   
7                                            Don Mills   
9                             Garden District, Ryerson   
11   West Deane Park, Princess Gardens, Martin Grov...   
12              Rouge Hill, Port Union, Highland Creek   
13                                           Don Mills   
14                                    Woodbine Heights   
15                                      St. James Town   
16                                  Humewood-Cedarvale   
17   Eringate, Bloordale Gardens, Old Burnhamthorpe...   
19                                         The Beaches   
20                                         Berczy Park   
23                                             Leaside   
24                                  Central Bay Street   
25                                            Christie   
26                                           Cedarbrae   
30                            Richmond, Adelaide, King   
33                        Fairview, Henry Farm, Oriole   
35                                        East Toronto   
36   Harbourfront East, Union Station, Toronto Islands   
37                            Little Portugal, Trinity   
38         Kennedy Park, Ionview, East Birchmount Park   
41                        The Danforth West, Riverdale   
42            Toronto Dominion Centre, Design Exchange   
43        Brockton, Parkdale Village, Exhibition Place   
44                     Golden Mile, Clairlea, Oakridge   
47                      India Bazaar, The Beaches West   
48                      Commerce Court, Victoria Hotel   
50                                       Humber Summit   
52                             Willowdale, Newtonbrook   
54                                     Studio District   
55                   Bedford Park, Lawrence Manor East   
57                                    Humberlea, Emery   
58                         Birch Cliff, Cliffside West   
59                                          Willowdale   
61                                       Lawrence Park   
63                       Runnymede, The Junction North   
65   Dorset Park, Wexford Heights, Scarborough Town...   
67                                    Davisville North   
68                            Forest Hill North & West   
69                       High Park, The Junction South   
71                                   Wexford, Maryvale   
72                                          Willowdale   
73                                  North Toronto West   
74                 The Annex, North Midtown, Yorkville   
75                              Parkdale, Roncesvalles   
76               Canada Post Gateway Processing Centre   
79                                          Davisville   
80                      University of Toronto, Harbord   
81                                  Runnymede, Swansea   
83                         Moore Park, Summerhill East   
84           Kensington Market, Chinatown, Grange Park   
86   Summerhill West, Rathnelly, South Hill, Forest...   
87   CN Tower, King and Spadina, Railway Lands, Har...   
88        New Toronto, Mimico South, Humber Bay Shores   
92                                      Stn A PO Boxes   
94                                           Northwest   
96                         St. James Town, Cabbagetown   
97              First Canadian Place, Underground city   
98      The Kingsway, Montgomery Road , Old Mill North   
99                                Church and Wellesley   
100              Business reply mail Processing CentrE   
102  Mimico NW, The Queensway West, South of Bloor,...   

         1st Most Common Venue   2nd Mo

#### Cluster 3

In [28]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue  \
101  Old Mill South, King's Mill Park, Sunnylea, Hu...                  Park   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
101        Baseball Field         Shopping Mall         Women's Store   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
101     Electronics Store               Dog Run      Doner Restaurant   

    8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
101            Donut Shop   Dumpling Restaurant  Eastern European Restaurant

#### Cluster 4

In [29]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue  \
21                       Caledonia-Fairbanks                  Park   
22                                    Woburn                  Park   
49  North Park, Maple Leaf Park, Upwood Park           Gas Station   
66                           York Mills West                  Park   
91                                  Rosedale                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
21         Women's Store   Japanese Restaurant            Beer Store   
22           Coffee Shop         Women's Store   Empanada Restaurant   
49                Bakery     Convenience Store                  Park   
66          Tennis Court                  Bank                   Gym   
91           Coffee Shop         Grocery Store            Playground   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
21    Mexican Restaurant   Sporting Goods Shop                Bakery   
22               Dog Run      Doner Restaurant            Donut Shop   
49           Coffee Shop     Electronics Store               Dog Run   
66             Pet Store     French Restaurant     Convenience Store   
91                 Trail           Candy Store        Sandwich Place   

   8th Most Common Venue        9th Most Common Venue 10th Most Common Venue  
21         Grocery Store                          Gym            Event Space  
22   Dumpling Restaurant  Eastern European Restaurant      Electronics Store  
49      Doner Restaurant                   Donut Shop    Dumpling Restaurant  
66            Restaurant                      Dog Run             Donut Shop  
91   Japanese Restaurant                         Bank     Athletics & Sports

#### Cluster 5

In [30]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
5   Islington Avenue              Pharmacy           Golf Course   
62          Roselawn            Playground          Skating Rink   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
5                   Park                  Café         Shopping Mall   
62                Garden                  Café              Pharmacy   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
5           Skating Rink                  Bank            Playground   
62    Falafel Restaurant           Event Space  Ethiopian Restaurant   

   9th Most Common Venue 10th Most Common Venue  
5          Grocery Store     Golf Driving Range  
62   Empanada Restaurant         Farmers Market

#### And finally our manually created cluster 6 with its only member: Upper Rouge

In [31]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 5, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
95  Upper Rouge                   NaN                   NaN   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
95                   NaN                   NaN                   NaN   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
95                   NaN                   NaN                   NaN   

   9th Most Common Venue 10th Most Common Venue  
95                   NaN                    NaN